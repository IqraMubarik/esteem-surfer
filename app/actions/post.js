// @flow
import {Client} from 'dsteem';

const client = new Client('https://api.steemit.com');

import type {postStateType, postActionType} from '../reducers/types';

/*
type actionType = {
    type: string,
    payload?: {},
};
*/

export const FETCH = 'FETCH';
export const FETCH_COMPLETE = 'FETCH_COMPLETE';
export const FETCH_ERROR = 'FETCH_ERROR';
export const SET_READ = 'SET_CONTENT_READ';
export const SET_VOTED = 'SET_CONTENT_VOTED';


export function fetchPosts(what, tag, startAuthor, startPermalink, limit = 20) {
    return (dispatch: (action: actionType) => void,
            getState: () => postStateType) => {
        const {content} = getState();

        dispatch({
            type: FETCH
        });

        client.database.getDiscussions().then((discussions) => {
            console.log(discussions)
        }).catch((err) => {
            dispatch({
                type: FETCH_ERROR
            });
        })
    };
}
