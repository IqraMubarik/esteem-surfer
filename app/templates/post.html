<nav-bar selected-section="'filters'"></nav-bar>
<div class="app-container keep-pos" id="content-main">
  <div class="app-content post-page">
    <div class="indicator post-indicator" ng-show="loadingPost">
      <i class="fa fa-spin fa-spinner fa-circle-o-notch"></i>
    </div>
    <div class="the-post" ng-if="!loadingPost" ng-class="{'fetching': loadingPost || loadingRest}">
      <div class="post-header">
        <h1 class="post-title">{{ title }}</h1>
        <div class="post-info-bar">
          <div class="post-author-pic" author-bg-img-style author="{{ post.author }}"></div>
          <div class="post-info-right-side">
            <span class="post-author"><author-name author-data="author"></author-name></span>
            <span class="post-author-reputation">{{ post.author_reputation|authorReputation|number:0 }}</span>
            <span class="post-cat"><span>{{ 'IN' | translate }}</span> <a ng-click="parentClicked()">{{ post.parent_permlink }}</a></span>
            <span class="post-date"><span
              title="{{ post.created|dateFormatted }}"> {{post.created|timeAgo}}</span></span>
          </div>
        </div>
        <div class="post-tools">
          <button class="btn btn-primary btn-bookmark" login-required on-login-success="addBookmark" uib-tooltip="{{ 'BOOKMARK' | translate }}"  ng-disabled="bookmarked">
            <i class="fa fa-bookmark-o" ng-if="!bookmarking && !bookmarked"></i>
            <i class="fa fa-bookmark" ng-if="bookmarked"></i>
            <i class="fa fa-spin fa-spinner fa-circle-o-notch" ng-show="bookmarking"></i>
          </button>
        </div>
      </div>
      <button class="btn btn-primary" ng-click="markdownHelperClicked()" ng-if="$root.showMarkdownResultHelper">Save
        Markdown Result to Test
      </button>
      <div class="post-body markdown-view selectable" ng-bind-html="body | markDown2Html"></div>
      <div class="post-footer">
        <div class="tag-list">
          <a ng-repeat="tag in tags" ng-click="tagClicked(tag)">{{ tag }}</a>
        </div>
        <div class="post-info-bar">
          <span class="post-date"><i class="fa fa-clock-o"></i> <span title="{{ post.created|dateFormatted }}"> {{post.created|timeAgo}}</span></span>
          <span class="post-author">{{ 'BY' | translate }} <author-name author-data="author"></author-name></span>
          <span class="post-author-reputation">{{ post.author_reputation|authorReputation|number:0 }}</span>
          <div class="post-interaction">
            <a class="post-resteem" ng-click="resteem()"><i class="fa fa-retweet"></i></a>
            <a class="post-leave-comment" ng-click="leaveComment()">Reply</a>
            <span class="post-comment-count"><i class="fa fa-comments"></i> {{ post.children }}</span>
          </div>
        </div>
        <div class="post-voting">
          <div class="post-up-vote">
            <a ng-click="upVoteClicked(post)">
              <i class="fa fa-chevron-circle-up"></i>
            </a>
          </div>
          <div class="post-total">
            <content-payout-info content="post"></content-payout-info>
          </div>
        </div>
        <div class="post-voters">
          <content-voters-info content="post"></content-voters-info>
        </div>
      </div>
      <div class="indicator comments-indicator" ng-show="loadingRest">
        <i class="fa fa-spin fa-spinner fa-circle-o-notch"></i>
      </div>
      <div class="post-comment-list" ng-if="!loadingRest && commentsLength">
        <div class="comment-list-header">
          <div class="list-pagination">
            <span class="page-label">page:</span>
            <span class="cur-page">{{ commentsCurPage }}</span><span
            class="total-pages">{{ commentsTotalPages }}</span>
            <a class="btn-prev" ng-class="{'disabled': !commentsHasPrev}" ng-click="commentsGoPrev()"><i
              class="fa fa-chevron-left"></i></a>
            <a class="btn-next" ng-class="{'disabled': !commentsHasNext}" ng-click="commentsGoNext()"><i
              class="fa fa-chevron-right"></i></a>
          </div>
        </div>
        <comment-list comments="comments"></comment-list>
        <div class="comment-list-footer">
          <div class="list-pagination">
            <span class="page-label">{{ 'COMMENTS_PAGE' | __ }}</span>
            <span class="cur-page">{{ commentsCurPage }}</span><span
            class="total-pages">{{ commentsTotalPages }}</span>
            <a class="btn-prev" ng-class="{'disabled': !commentsHasPrev}" ng-click="commentsGoPrev()"><i
              class="fa fa-chevron-left"></i></a>
            <a class="btn-next" ng-class="{'disabled': !commentsHasNext}" ng-click="commentsGoNext()"><i
              class="fa fa-chevron-right"></i></a>
          </div>
        </div>
      </div>
      <div class="post-comment-list-empty" ng-if="!loadingRest && !commentsLength">
        {{ 'NO_COMMENTS_YET' | __ }}
      </div>
    </div>
  </div>
</div>
<app-footer></app-footer>
