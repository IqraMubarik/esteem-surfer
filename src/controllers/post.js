import {authorReputation} from '../filters/author-reputation';

export default ($scope, $rootScope, $routeParams, $filter, $timeout, $uibModal, $location, $q, steemService, eSteemService, helperService, activeUsername, activePostFilter, constants) => {

  let parent = $routeParams.parent;
  let author = $routeParams.author;
  let permlink = $routeParams.permlink;
  let commentId = parseInt($routeParams.comment);

  let routePath = `/${parent}/@${author}/${permlink}`;
  let contentPath = `${author}/${permlink}`;

  let pathData = {};

  let commentsData = [];

  const commentsPerPage = constants.commentListSize;

  $scope.commentsLength = 0; // Length of compiled comments array (Not total comments count)
  $scope.commentsHasPrev = false;
  $scope.commentsHasNext = false;
  $scope.commentsCurPage = 1;
  $scope.commentsTotalPages = 0;

  $scope.commentSorting = {
    model: 'trending',
    opts: [
      {value: 'trending', name: $filter('__')('COMMENT_SORT_ORDER_TRENDING')},
      {value: 'author_reputation', name: $filter('__')('COMMENT_SORT_ORDER_AUTHOR_REPUTATION')},
      {value: 'votes', name: $filter('__')('COMMENT_SORT_ORDER_VOTES')},
      {value: 'created', name: $filter('__')('COMMENT_SORT_ORDER_CREATED')}
    ]
  };

  $scope.commentSortFieldChanged = () => {
    // Rebuild comment list
    let content = pathData.content[contentPath];
    commentsData = compileComments(content);
    $scope.commentsLength = commentsData.length;
    $scope.commentsTotalPages = Math.ceil(commentsData.length / commentsPerPage);
    $scope.sliceComments();
  };

  const compileComments = (parent) => {
    const sortOrders = {
      trending: (a, b) => {
        const keyA = a['pending_payout_value'],
          keyB = b['pending_payout_value'];

        if (keyA > keyB) return -1;
        if (keyA < keyB) return 1;
        return 0;
      },
      author_reputation: (a, b) => {
        const keyA = authorReputation(a['author_reputation']),
          keyB = authorReputation(b['author_reputation']);

        if (keyA > keyB) return -1;
        if (keyA < keyB) return 1;
        return 0;
      },
      votes: (a, b) => {
        const keyA = a['net_votes'],
          keyB = b['net_votes'];

        if (keyA > keyB) return -1;
        if (keyA < keyB) return 1;
        return 0;
      },
      created: (a, b) => {
        const keyA = Date.parse(a['created']),
          keyB = Date.parse(b['created']);

        if (keyA > keyB) return -1;
        if (keyA < keyB) return 1;
        return 0;
      }
    };

    let comments = [];
    for (let k of parent.replies) {

      let reply = pathData.content[k];

      comments.push(
        Object.assign(
          {},
          reply,
          {comments: compileComments(reply)},
          {author_data: pathData.accounts[reply.author]},
          {_selected_: reply.id === commentId}
        )
      )
    }

    comments.sort(sortOrders[$scope.commentSorting.model]);

    return comments
  };

  const loadState = () => {
    return steemService.getState(routePath).then(stateData => {
      pathData = stateData;

      let content = stateData.content[contentPath];
      $scope.post = content;

      $rootScope.refreshContent(content);

      $scope.author = pathData.accounts[author];

      commentsData = compileComments(content);
      $scope.commentsLength = commentsData.length;
      $scope.commentsTotalPages = Math.ceil(commentsData.length / commentsPerPage);
      $scope.sliceComments();

      if (commentId) {
        // If comment specified (coming from author detail page comments or replies)
        // find out comment row and move pagination according to it's place.

        let i = 1;
        let commentPlace = null;
        for (let k of commentsData) {
          if (k.id === commentId) {
            commentPlace = i;
            break;
          } else {
            // TODO: Find better approach like recursive lookup.
            let j = JSON.stringify(k);
            if (j.match(new RegExp(`"id": ?${commentId},`))) {
              commentPlace = i;
              break;
            }
          }
          i += 1;
        }

        if (commentPlace) {
          // Find out which page comment is on and set forward the pagination
          let commentPage = Math.ceil(commentPlace / commentsPerPage);
          if (commentPage > 1) {
            $scope.commentsGoPage(commentPage);
          }

          // Scroll page to selected comment
          $timeout(() => {
            scrollToSelectedComment();
          }, 500)
        }

      }
    })
  };

  const scrollToComments = () => {
    let e = document.querySelector('#content-main');
    e.scrollTop = e.scrollHeight;
  };

  const scrollToSelectedComment = () => {
    let e = document.querySelector('#content-main');
    let c = document.querySelector('.comment-list-item.selected');
    e.scrollTop = (c.offsetTop - 200);
  };

  $scope.commentsGoPage = (page) => {
    $scope.commentsCurPage = page;
    $scope.sliceComments();
  };

  $scope.commentsGoNext = () => {
    $scope.commentsCurPage += 1;
    $scope.sliceComments();
  };

  $scope.commentsGoPrev = () => {
    $scope.commentsCurPage -= 1;
    $scope.sliceComments();
  };

  $scope.sliceComments = () => {
    let start = ( $scope.commentsCurPage - 1) * commentsPerPage;
    let end = start + commentsPerPage;

    $scope.comments = commentsData.slice(start, end);

    $scope.commentsHasPrev = $scope.commentsCurPage > 1;
    $scope.commentsHasNext = $scope.commentsCurPage < $scope.commentsTotalPages;
  };

  $scope.post = $rootScope.Data['post'] || null;

  let init = () => {
    let defer = $q.defer();

    if ($scope.post) {
      defer.resolve($scope.post);
    } else if ($rootScope.selectedPost && $rootScope.selectedPost.permlink === permlink && $rootScope.selectedPost.author === author) {
      // The last selected post from list === this post
      defer.resolve($rootScope.selectedPost);
    } else {
      // When clicked outside from posts lists (post body) or refreshed in development environment
      steemService.getContent(author, permlink).then((resp) => {
        defer.resolve(resp);
      }).catch((e) => {
        defer.reject(e)
      })
    }
    return defer.promise;
  };

  const main = () => {
    $scope.loadingPost = true;
    $scope.loadingRest = true;

    init().then((content) => {
      $scope.post = content;

      // Defined separately because $scope.post will be changed after state loaded.
      $scope.title = content.title;
      $scope.body = content.body;

      let jsonMeta = {};
      try {
        jsonMeta = JSON.parse($scope.post.json_metadata);
      } catch (e) {
      }

      $scope.app = $filter('appName')(jsonMeta.app);

      // Sometimes tag list comes with duplicate items. Needs to singularize.
      $scope.tags = [...new Set(jsonMeta.tags)];

      // Temporary author data while loading original in background
      $scope.author = {name: $scope.post.author};

      $scope.loadingPost = false;

      const archived = content.cashout_time === '1969-12-31T23:59:59';

      $scope.canEdit = content.author === activeUsername() && !archived;

      loadState().catch((e) => {
        // TODO: Handle catch
      }).then(() => {
        $scope.loadingRest = false;
      });

      // Mark post as read
      helperService.setPostRead(content.author, content.permlink);

      // Add to nav history
      $rootScope.setNavVar('post', content);

      // Check if post bookmarked
      for (let i of $rootScope.bookmarks) {
        if (i.permlink === $scope.post.permlink) {
          $scope.bookmarked = true;
          break;
        }
      }

      if (commentId) {
        // Scroll page to comments section
        $timeout(() => {
          scrollToComments();
        }, 800)
      }
    });
  };

  main();

  $scope.parentClicked = () => {
    let u = `/posts/${activePostFilter()}/${$scope.post.parent_permlink}`;
    $location.path(u);
  };

  $scope.tagClicked = (tag) => {
    let u = `/posts/${activePostFilter()}/${tag}`;
    $location.path(u);
  };

  $scope.addBookmark = () => {
    if ($scope.bookmarked) {
      return false;
    }

    $scope.bookmarking = true;
    $scope.bookmarked = false;
    eSteemService.addBookmark(activeUsername(), $scope.post).then((resp) => {
      $scope.bookmarked = true;
      $rootScope.$broadcast('newBookmark', {id: $scope.post.id});
    }).catch((e) => {
      $rootScope.showError('Could not added to bookmarks!')
    }).then(() => {
      $scope.bookmarking = false;
    });
  };

  $scope.markdownHelperClicked = () => {
    let m = $scope.post.body;
    $rootScope.saveMarkdownResult($scope.post.id, m);
  };

  $scope.commentId = commentId;

  /*
  $scope.replyClicked = () => {
    $rootScope.openReplyWindow($scope.post, (resp) => {
      $scope.comments.push(resp);
      $rootScope.closeReplyWindow();
    })
  }*/

  $scope.commentFlag = false;

  $scope.replyClicked = () => {
    $rootScope.$broadcast('commentEditorOpening');
    $timeout(() => {
      $scope.commentFlag = true;
    }, 100);
  };

  $scope.editClicked = () => {
    const u = `/editor/${$scope.post.author}/${$scope.post.permlink}`;
    $location.path(u);
  };

  $rootScope.$on('commentEditorOpening', () => {
    $scope.commentFlag = false;
  });

  $scope.onCommentEditorCanceled = () => {
    $scope.commentFlag = false;
  };

  $scope.afterNewComment = (newComment) => {
    if ($scope.commentsLength === 0) {
      commentsData = [newComment];
      $scope.commentsLength = commentsData.length;
      $scope.commentsTotalPages = Math.ceil(commentsData.length / commentsPerPage);
      $scope.sliceComments();
    } else {
      $scope.comments.push(newComment);
    }
  };

  $scope.refresh = () => {
    $scope.post = null;
    $rootScope.selectedPost = null;
    main();
  }
};
