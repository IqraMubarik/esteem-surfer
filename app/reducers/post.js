// @flow

import {FETCH, FETCH_COMPLETE, FETCH_ERROR, SET_READ, SET_VOTED} from '../actions/post';

import type {postActionType} from './types';

/*
export type actionType = {
    type: string,
    payload?: {},
};
*/

const defaultState = {
    posts: {},
    groups: {}
};


export default function posts(state: {} = defaultState, action: postActionType) {

    let newState, groupKey;

    switch (action.type) {
        case FETCH:

            newState = JSON.parse(JSON.stringify(state));
            groupKey = action.payload.group;

            if (newState.groups[groupKey] === undefined) {
                newState.groups[groupKey] = {
                    ids: [],
                    loading: true
                }
            }

            newState.groups[groupKey].loading = true;

            return newState;
        case FETCH_COMPLETE:

            newState = JSON.parse(JSON.stringify(state));
            groupKey = action.payload.group;

            if (newState.groups[groupKey] === undefined) {
                newState.groups[groupKey] = {
                    ids: [],
                    loading: true
                }
            }

            const data = action.payload.data;

            newState.groups[groupKey].ids = data.map(i => i.id);
            newState.groups[groupKey].loading = false;

            for (let r of data) {
                newState.posts[r.id] = r;
            }

            return newState;
        default:
            return state;
    }
}

