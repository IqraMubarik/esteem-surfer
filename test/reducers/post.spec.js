import post from '../../app/reducers/post';
import {
    FETCH, FETCH_COMPLETE

} from '../../app/actions/post';

import deepFreeze from 'deep-freeze';


describe('reducers', () => {
    describe('post', () => {


        it('should handle initial state', () => {
            expect(post(undefined, {})).toMatchSnapshot();
        });

        it('should handle FETCH', () => {
            const stateBefore = {
                posts: {},
                groups: {}
            };

            deepFreeze(stateBefore);

            const action = {
                type: FETCH,
                payload: {
                    group: 'trending'
                }
            };

            const stateAfter = {
                posts: {},
                groups: {
                    trending: {
                        ids: [],
                        loading: true
                    }
                }
            };

            const res = post(stateBefore, action);

            expect(res).toEqual(stateAfter);
        });

        it('should handle FETCH_COMPLETE', () => {
            const stateBefore = {
                posts: {},
                groups: {
                    trending: {
                        ids: [],
                        loading: true
                    }
                }
            };

            deepFreeze(stateBefore);

            const action = {
                type: FETCH_COMPLETE,
                payload: {
                    data: [
                        {id: 12, author: 'good-karma', permlink: 'welcome to esteem', votes: 5},
                        {id: 13, author: 'talhasch', permlink: 'lorem ipsum', votes: 12},
                        {id: 14, author: 'chrisbolten', permlink: 'sit amet', votes: 3}
                    ],
                    group: 'trending'
                }
            };

            const stateAfter = {
                posts: {
                    12: {id: 12, author: 'good-karma', permlink: 'welcome to esteem', votes: 5},
                    13: {id: 13, author: 'talhasch', permlink: 'lorem ipsum', votes: 12},
                    14: {id: 14, author: 'chrisbolten', permlink: 'sit amet', votes: 3}
                },
                groups: {
                    trending: {
                        ids: [12, 13, 14],
                        loading: false
                    }
                }
            };

            const res = post(stateBefore, action);

            expect(res).toEqual(stateAfter);
        });


        it('should handle FETCH_COMPLETE', () => {
            const stateBefore = {
                posts: {
                    12: {id: 12, author: 'good-karma', permlink: 'welcome to esteem', votes: 5},
                    13: {id: 13, author: 'talhasch', permlink: 'lorem ipsum', votes: 12},
                    14: {id: 14, author: 'chrisbolten', permlink: 'sit amet', votes: 3}
                },
                groups: {
                    trending: {
                        ids: [12, 13, 14],
                        loading: false
                    }
                }
            };

            deepFreeze(stateBefore);

            const action = {
                type: FETCH_COMPLETE,
                payload: {
                    data: [
                        {id: 18, author: 'foo', permlink: 'lorem ipsum', votes: 222},
                        {id: 12, author: 'good-karma', permlink: 'welcome to esteem', votes: 21},
                    ],
                    group: 'active'
                }
            };

            const stateAfter = {
                posts: {
                    12: {id: 12, author: 'good-karma', permlink: 'welcome to esteem', votes: 21},
                    13: {id: 13, author: 'talhasch', permlink: 'lorem ipsum', votes: 12},
                    14: {id: 14, author: 'chrisbolten', permlink: 'sit amet', votes: 3},
                    18: {id: 18, author: 'foo', permlink: 'lorem ipsum', votes: 222}
                },
                groups: {
                    trending: {
                        ids: [12, 13, 14],
                        loading: false
                    },
                    active: {
                        ids: [18, 12],
                        loading: false
                    }
                }
            };

            const res = post(stateBefore, action);

            expect(res).toEqual(stateAfter);
        });
    });
});
