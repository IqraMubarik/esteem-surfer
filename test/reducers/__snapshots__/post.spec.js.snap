// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers post should handle initial state 1`] = `
Object {
  "groups": Object {},
  "posts": Object {},
}
`;
